File: framework/src/main/java/com/thornbirds/frameworkext/component/presenter/RxLifeCyclePresenter.java
package com.thornbirds.frameworkext.component.presenter;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;

import com.thornbirds.frameworkext.component.page.BasePageController;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.Predicate;
import io.reactivex.subjects.BehaviorSubject;

/**
 * Created by yangli on 2017/8/2.
 *
 * @module 具有Rx生命周期绑定的表现
 */
public abstract class RxLifeCyclePresenter<VIEW> extends ComponentPresenter<VIEW> {

    protected final BehaviorSubject<PresenterEvent> mBehaviorSubject = BehaviorSubject.create();

    public RxLifeCyclePresenter(BasePageController controller) {
        super(controller);
    }

    @Override
    @CallSuper
    public void startPresenter() {
        mBehaviorSubject.onNext(PresenterEvent.START);
    }

    @Override
    @CallSuper
    public void stopPresenter() {
        mBehaviorSubject.onNext(PresenterEvent.STOP);
    }

    @Override
    @CallSuper
    public void destroy() {
        mBehaviorSubject.onNext(PresenterEvent.DESTROY);
    }

    protected final <T> ObservableTransformer<T, T> bindUntilEvent(PresenterEvent event) {
        return bindUntilEvent(mBehaviorSubject, event);
    }

    private static <T, R> ObservableTransformer<T, T> bindUntilEvent(
            final BehaviorSubject<R> subject, final R event) {
        return new ObservableTransformer<T, T>() {
            @Override
            public ObservableSource<T> apply(@NonNull Observable<T> upstream) {
                return upstream.takeUntil(
                        subject.filter(new Predicate<R>() {
                            @Override
                            public boolean test(@NonNull R r) throws Exception {
                                return r == event;
                            }
                        })).take(1);
            }
        };
    }

    protected enum PresenterEvent {
        START, STOP, DESTROY
    }
}

File: framework/src/main/java/com/thornbirds/frameworkext/component/presenter/RxLifeCyclePresenter.kt
package com.thornbirds.frameworkext.component.presenter

import androidx.annotation.CallSuper
import com.thornbirds.frameworkext.component.page.BasePageController
import io.reactivex.ObservableTransformer
import io.reactivex.subjects.BehaviorSubject
import com.thornbirds.component.presenter.ComponentPresenter as Presenter

/**
 * Created by yangli on 2017/8/2.
 *
 * @module 具有Rx生命周期绑定的表现
 */
typealias ComponentPresenter<VIEW> = Presenter<VIEW, BasePageController>

abstract class RxLifeCyclePresenter<VIEW>(
    controller: BasePageController
) : ComponentPresenter<VIEW>(controller) {
    private val mBehaviorSubject: BehaviorSubject<PresenterEvent> = BehaviorSubject.create()

    @CallSuper
    override fun startPresenter() {
        mBehaviorSubject.onNext(PresenterEvent.START)
    }

    @CallSuper
    override fun stopPresenter() {
        mBehaviorSubject.onNext(PresenterEvent.STOP)
    }

    @CallSuper
    override fun destroy() {
        mBehaviorSubject.onNext(PresenterEvent.DESTROY)
    }

    protected fun <T> bindUntilEvent(event: PresenterEvent): ObservableTransformer<T, T> {
        return bindUntilEvent(mBehaviorSubject, event)
    }

    protected enum class PresenterEvent {
        START, STOP, DESTROY
    }

    companion object {
        private fun <T, R> bindUntilEvent(
            subject: BehaviorSubject<R>, event: R
        ): ObservableTransformer<T, T> {
            return ObservableTransformer { upstream ->
                upstream.takeUntil(
                    subject.filter { r -> r == event }).take(1)
            }
        }
    }
}